[ROCM] query warp size for host code, do not use C10_WARP_SIZE

micmelesse opened the pull request:
ROCm supports gfx targets with 32 and 64 warp size. Device compilation
correctly handles the C10_WARP_SIZE (aka warpSize) constant. Host
compilation cannot rely on a single hard-coded value, but instead needs
to query device properties at runtime.

worked with @jeffdaily 


cc @jeffdaily @sunway513 @jithunnair-amd @ROCmSupport @KyleCZH

micmelesse committed:
Rocm5.0 internal testing warpsize (#870)

add warpsize fixes to kernels

micmelesse committed:
fix exec issues

micmelesse committed:
fix aten/src/ATen/native/cuda/Embedding.cu

micmelesse committed:
replace THCCeilDiv with ceil_div

micmelesse committed:
fix ATenCeilDiv

micmelesse committed:
fix more unkown op issues

micmelesse committed:
Rocm5.0 internal testing binomial bug pr (#871)

fix distribution_binary_kernel

zasdfgbnm commented on aten/src/ATen/native/cuda/DistributionTemplates.h:
```diff
@@ -254,6 +254,9 @@ void distribution_binary_kernel(TensorIterator &iter, PhiloxCudaState philox_arg
   const input_t_1 *input_data_1 = static_cast<const input_t_1 *>(iter.data_ptr(1));
   const input_t_2 *input_data_2 = static_cast<const input_t_2 *>(iter.data_ptr(2));
 
+  const int num_threads = at::cuda::warp_size() * 2;
```
In https://github.com/pytorch/pytorch/pull/68008, this value is changed to warpsize * 4

zasdfgbnm commented on aten/src/ATen/native/cuda/Embedding.cu:
```diff
@@ -315,18 +316,20 @@ Tensor & embedding_renorm_cuda_(Tensor & self, const Tensor & indices,
-                  "BlockReduceSum requires all warps be active");
+    int warp_size = at::cuda::warp_size();
+    int max_threads = warp_size * warp_size;
+    const int num_threads = warp_size * 2;
```
Why this is changed from 128 to warp_size * 2?

micmelesse replied:
reverted.

zasdfgbnm commented on aten/src/ATen/native/cuda/FunctionOfAMatrixUtilsKernel.cu:
```diff
@@ -24,19 +24,26 @@ __global__ void _elemwise_kernel(int total_n_elems, func_t f) {
     total_n_elems >= 0 && total_n_elems <= std::numeric_limits<int32_t>::max()
   );
 
+  const int n_threads = at::cuda::warp_size() * 2;
```
In https://github.com/pytorch/pytorch/pull/68008, this value is changed to warpsize * 4

zasdfgbnm commented on aten/src/ATen/native/cuda/LinearAlgebra.cu:
```diff
@@ -84,19 +84,26 @@ __global__ void _elementwise_kernel(int total_n_elems, func_t f) {
+  if (n_threads == 64) {
+    _elementwise_kernel<64, n_elems_per_thread, func_t>
+        <<<grid, block, 0, stream>>>(total_n_elems, f);
+  } else if (n_threads == 128) {
+    _elementwise_kernel<128, n_elems_per_thread, func_t>
+        <<<grid, block, 0, stream>>>(total_n_elems, f);
+  }
```
Should we guard this with ifdefs on ROCm? This increases the CUDA binary size.

peterbell10 replied:
If `C10_WARP_SIZE` works correctly for device code, then perhaps template parameter could be changed to `n_warps` and the kernel can calculate `n_threads` internally? That would avoid ROCm compiling two kernels for the same GPU architecture as well.

ngimel replied:
Elementwise kernel doesn't (or shouldn't, I didn't check) have warp-specialized code, so for both ROCm and cuda it should be compiled with 128 threads.

micmelesse replied:
see comment on scattergatherkernel.cu

zasdfgbnm commented on aten/src/ATen/native/cuda/ScatterGatherKernel.cu:
```diff
@@ -65,17 +65,27 @@ __global__ void _scatter_gather_elementwise_kernel(int N, func_t f) {
+  if (nt == 64) {
+    _scatter_gather_elementwise_kernel<64, vt, func_t>
+        <<<grid, block, 0, stream>>>(N, f);
+  } else if (nt == 128) {
+    _scatter_gather_elementwise_kernel<128, vt, func_t>
+        <<<grid, block, 0, stream>>>(N, f);
+  }
+
```
Should we guard this with ifdefs on ROCm? This increases the CUDA binary size.

jeffdaily replied:
Call/launch sites are precisely the problem this PR is trying to address.  Launch sites must become dynamic and query the warp size at runtime.  `at::cuda::warp_size()`.  `C10_WARP_SIZE` is 32 for CUDA, and up to this point has been 64 for HIP.  For HIP, we must now support 32 and 64 depending on the gfx target.  The `C10_WARP_SIZE` for HIP was recently changed from 64 to `warpSize` that comes from HIP headers.  This does correct the warpSize used during compilation of kernels, but does not address the issue of querying the hardware at runtime.  We need a `num_threads()` that is a constexpr for compilation, but we need a dynamic `query_num_threads()` or `runtime_num_threads()` that perhaps wraps `at::cuda::warp_size() * 4`.

ngimel replied:
Do you actually want different number of threads depending on the warp size? Most call sites are for elementwise kernels that don't care about relationship to warp size, we specify num_threads() as 4*C10_WARP_SIZE, but what we really want is 128. Should we just modify `num_threads()` to not depend on the warp size?

micmelesse replied:
@ngimel If it is acceptable to have num_threads not rely on the warps size. Then that works for us.

micmelesse replied:
We have ifdef these changes so that they only affect rocm. It has a minor impact on our stack and we will address it in a subsequent pr. We will create an issue to track it.

zasdfgbnm commented on aten/src/ATen/native/cuda/UnfoldBackwardKernel.cu:
```diff
@@ -32,19 +32,26 @@ __global__ void _unfold_backward_elementwise_kernel(int total_n_elems, func_t f)
     total_n_elems >= 0 && total_n_elems <= std::numeric_limits<int32_t>::max()
   );
 
+  const int n_threads = at::cuda::warp_size() * 2;
```
In #68008, this value is changed to warpsize * 4

zasdfgbnm commented on aten/src/ATen/native/cuda/UnfoldBackwardKernel.cu:
```diff
@@ -32,19 +32,26 @@ __global__ void _unfold_backward_elementwise_kernel(int total_n_elems, func_t f)
+  if (n_threads == 64) {
+    _unfold_backward_elementwise_kernel<64, n_elems_per_thread, func_t>
+        <<<grid, block, 0, stream>>>(total_n_elems, f);
+  } else if (n_threads == 128) {
+    _unfold_backward_elementwise_kernel<128, n_elems_per_thread, func_t>
+        <<<grid, block, 0, stream>>>(total_n_elems, f);
+  }
```
Should we guard this with ifdefs on ROCm? This increases the CUDA binary size.

peterbell10 commented on aten/src/ATen/native/cuda/TensorModeKernel.cu:
```diff
@@ -139,8 +139,8 @@ void handle_fused_mode(
-  static_assert(num_threads % C10_WARP_SIZE == 0 &&
-                num_threads <= cuda_utils::kCUDABlockReduceMaxThreads, "");
+  TORCH_INTERNAL_ASSERT(num_threads % at::cuda::warp_size() == 0);
+  static_assert(num_threads <= cuda_utils::kCUDABlockReduceMaxThreads, "");
```
How does `kCUDABlockReduceMaxThreads` work here when it's defined in terms of `C10_WARP_SIZE`?
https://github.com/pytorch/pytorch/blob/f9422e1c6b5dab15eacbbb0e1d4370b07db1e8e8/aten/src/ATen/native/cuda/block_reduce.cuh#L17

micmelesse replied:
fixed

micmelesse committed:
Merge branch 'master' into fix_warpsize_issue

micmelesse committed:
refactor with new num_threads function

micmelesse committed:
fix trailing white space

micmelesse committed:
Merge branch 'master' into fix_warpsize_issue

micmelesse committed:
revert num_threads change in embedding.cu

micmelesse committed:
use constexpr n_threads for cuda

micmelesse committed:
make nt is warp*4 for template kernels

micmelesse committed:
remove kernel assert changes

micmelesse commented:
Hi, @zasdfgbnm @peterbell10 @ngimel 

Can we get a rereview? We have addressed most of the comments. We have decided to ifdef the changes that involve launching elementwise kernels so that they only affect rocm. It has a very minor impact on our stack and we will address it in a subsequent pr. This pr is important to an internal project and already quite big. We will create an issue to track the issue. 

We have discussed this with @malfet in our last engineering sync.

micmelesse committed:
Merge branch 'master' into fix_warpsize_issue

micmelesse commented:
@malfet Please see https://github.com/pytorch/pytorch/pull/69942
